analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,171 @@
+                {
+                    var additionalLocations = property.RouteParamLocation is not null
+                        ? new[] { property.RouteParamLocation }
+                        : Array.Empty<Location>();
+                    var message = property.RouteParamLocation is not null
+                        ? string.Format(MessageWithSecondaryLocationFormat, property.Type.GetName())
+                        : string.Format(MessageNoSecondaryLocationFormat, property.Type.GetName(), property.ConstraintType, property.Route);
+
+                    c.ReportIssue(Diagnostic.Create(Rule,
+                        property.Type.GetLocation(),
+                        additionalLocations,
+                        additionalLocations.ToProperties(string.Format(SecondaryMessageFormat, property.ConstraintType)),
+                        message));
+                }
+            },
+            SyntaxKind.ClassDeclaration);
+        });
+
+    private static IList<PropertyTypeMismatch> GetPropertyTypeMismatches(ClassDeclarationSyntax classDeclaration, SemanticModel semanticModel)
+    {
+        var routeParameters = GetRouteParametersWithValidConstraint(classDeclaration);
+
+        if (routeParameters.Count == 0)
+        {
+            return Array.Empty<PropertyTypeMismatch>();
+        }
+
+        var mismatches = new List<PropertyTypeMismatch>();
+        var classSymbol = semanticModel.GetDeclaredSymbol(classDeclaration);
+        var properties = classSymbol.GetMembers()
+            .Where(member => member.Kind is SymbolKind.Property && member.HasAttribute(KnownType.Microsoft_AspNetCore_Components_ParameterAttribute) && routeParameters.ContainsKey(member.Name));
+
+        foreach (var property in properties)
+        {
+            foreach (var routeParameter in routeParameters[property.Name])
+            {
+                if (TypeMatchConstraint(property.GetSymbolType(), routeParameter.Constraint))
+                {
+                    continue;
+                }
+
+                foreach (var propertySyntax in property.DeclaringSyntaxReferences.Where(r => r.GetSyntax() is PropertyDeclarationSyntax).Select(r => (PropertyDeclarationSyntax)r.GetSyntax()))
+                {
+                    mismatches.Add(new PropertyTypeMismatch(propertySyntax.Type, routeParameter.Constraint, routeParameter.FromRoute, routeParameter.RouteParamLocation));
+                }
+            }
+        }
+
+        return mismatches;
+    }
+
+    private static bool TypeMatchConstraint(ISymbol type, string constraintType) =>
+        ConstraintMapping.ContainsKey(constraintType) && ConstraintMapping[constraintType] == type.Name;
+
+    private static Dictionary<string, List<RouteParameter>> GetRouteParametersWithValidConstraint(ClassDeclarationSyntax classDeclaration)
+    {
+        var routeParameters = new Dictionary<string, List<RouteParameter>>(StringComparer.InvariantCultureIgnoreCase);
```

Comment:
Yes you can, everything is case-insensitive in the URL.
Like `/route/{BoolParam:bool}` is the same as `/route/{BOOLparam:bOOl}`

I am adding test cases for this since I forgot to add some.